ENT-1909: First iteration of RngEnclavelet

exFalso opened the pull request:
This PR is the first iteration of a simple RNG enclave showing how to build a JVM enclave. The enclave generates random bytes and signs over it with a key rooted in an attested quote.

Note that the test is using Simulation libraries, so the quote isn't generated/checked. Followup task is to add tests that show how to do this, given sufficient hardware/attestation capabilities.

The enclave implements the `EnclaveletEnclave` class, which means it's compatible with the enclavelet host infrastructure.

IgorNitto commented on build.gradle:
```diff
@@ -0,0 +1,40 @@
+dependencies {
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk8:$kotlin_version"
+    implementation "com.r3.sgx:api-core-enclave:$oblivium_version"
+    implementation "org.bouncycastle:bcprov-jdk15on:$bouncycastle_version"
```
Where is bouncycastle required?

exFalso replied:
Removed

IgorNitto commented on src/main/kotlin/com/r3/sgx/enclave/rng/RngEnclave.kt:
```diff
@@ -0,0 +1,52 @@
+import kotlin.math.min
+
+class RngEnclave : EnclaveletEnclave() {
+    private lateinit var signatureScheme: SignatureScheme
```
Outside the scope of these changes: `EnclaveletEnclave` is not a very good name IMHO. If this is the main facade into the framework, why not just calling it `Enclave`? A more specific terminology would also be ok, but adding the word `Enclavelet` does not convey any additional meaning to me about this class or interface.

mikehearn replied:
Enclavelet is fine, it's a word we can 'own' :)

exFalso replied:
Sweet, changed it

IgorNitto commented on build.gradle:
```diff
@@ -0,0 +1,40 @@
+buildscript {
+    repositories {
+        jcenter()
```
I tried compiling it but gradle cannot find the classpath dependencies using these repositories settings

exFalso replied:
It should work now, i published some nightly artifacts manually that should work. But this isn't a requirement, once automatic publishing is set up i'll add CI for this repo.

In the meantime if you want to try compiling it you need composite builds. To use that add a `settings.gradle` like so:
```
includeBuild('oblivium/sgx-jvm-plugin')
includeBuild('oblivium')
```
And add a symlink to a local checkout of oblivium named `oblivium` next to `settings.gradle`. You'll need the `ENT-1909/aslemmer-rng-enclavelet` branch for it to compile

mikehearn commented on src/main/kotlin/com/r3/sgx/enclave/rng/RngEnclave.kt:
```diff
@@ -0,0 +1,52 @@
+import java.security.MessageDigest
+import kotlin.math.min
+
+class RngEnclave : EnclaveletEnclave() {
```
This is a demo class, so it needs a LOT of detailed comments! Imagine you're explaining this to a colleague who never saw your API before and is watching over your shoulder as you narrate it.

exFalso replied:
Documentation is a followup task (ENT-2631), I'd like to get build/CI out of the way first.

mikehearn replied:
Generally speaking we prefer to have docs in the same PR as the code, not as a followup. That way the reviewers can ensure the docs actually match the code, the docs explain things that otherwise reviewers would have to ask or puzzle out, and we don't get into a habit of forgetting. It can easily be added to this branch.

exFalso replied:
Added a bunch of comments and a README. It's not really complete, I'd like to set up rst documentation similar to the corda one, referring to pieces of code.

IgorNitto commented on src/test/kotlin/com/r3/sgx/enclave/rng/RngEnclaveTest.kt:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    @Test
+    fun rngEnclaveWorks() {
```
Not a fundamental point but perhaps this code belong to a demo sample program more than a test, with some more comments explaining what is going on.

exFalso replied:
Added a bunch of comments here.

I see your point wrt a sample program. I was thinking of doing this in the next phase, providing a demo enclave host client that communicates with such an rng enclave

exFalso commented:
@mikehearn @IgorNitto ready for another round

mikehearn commented on src/test/kotlin/com/r3/sgx/enclave/rng/RngEnclaveTest.kt:
```diff
@@ -0,0 +1,111 @@
+        // A channel handler that records the binary blobs sent by the enclave.
+        val handler = BytesRecordingHandler()
+
+        // Configure the host side of attestation. Insert your SPID/quote type if you want to test using a
```
SPID? I thought that's a typo for EPID at first but it's used again below - what is a SPID? Where do I get one of these things from?

exFalso replied:
An SPID is given when one whitelists an IAS client key. I think this is what is used to identify Intel customers.

mikehearn replied:
Can that be explained in the code please?

exFalso replied:
I'll add a comment. If we decide to release our key we can also just burn that in.

mikehearn commented on src/test/kotlin/com/r3/sgx/enclave/rng/RngEnclaveTest.kt:
```diff
@@ -0,0 +1,111 @@
+
+        // Construct the host side of an Enclavelet, with channel and EPID attestation support
+        val connected = enclave.connected
+        val channels = connected.addDownstream(ChannelInitiatingHandler())
```
explicit types would help me understand what channels, connected and attesting actually are here.

exFalso replied:
Added more explanation and explicit types

mikehearn commented on src/test/kotlin/com/r3/sgx/enclave/rng/RngEnclaveTest.kt:
```diff
@@ -0,0 +1,111 @@
+
+        // Create a quote, including the enclave's report data created by RngEnclave.createReportData, and signed by the
+        // Quoting Enclave. Note that this test does *not* do a full attestation roundtrip to the Intel Attestation
+        // Service!
```
What's the significance of not doing the trip back to Intel HQ?

exFalso replied:
Added more on this

mikehearn commented on src/test/kotlin/com/r3/sgx/enclave/rng/RngEnclaveTest.kt:
```diff
@@ -0,0 +1,111 @@
+        // Quoting Enclave. Note that this test does *not* do a full attestation roundtrip to the Intel Attestation
+        // Service!
+        val quote = attesting.getQuote()
+        val hashedEnclaveKey = quote[quote.type.quote][SgxQuote.reportBody][SgxReportBody.reportData].read()
```
?? Is this a JSON structure or something? More types would help here.

exFalso replied:
I added an explanation of Cursors and specified the types explicitly. It's a bit verbose, but hopefully understandable.

mikehearn commented on src/test/kotlin/com/r3/sgx/enclave/rng/RngEnclaveTest.kt:
```diff
@@ -0,0 +1,111 @@
+        // HOST:    ocall_host_side    -- internals of receive in the host
+        // HOST:    handler.receive    -- the receive function in the handler we created at the beginning of the test
+        //
+        // Therefore when the above send() returns we will have already received the reply, recorded in handler.ocalls.
```
Great explanation!

mikehearn commented on src/test/kotlin/com/r3/sgx/enclave/rng/RngEnclaveTest.kt:
```diff
@@ -0,0 +1,111 @@
+        val keyDigest = MessageDigest.getInstance("SHA-512").digest(publicKey)
+        assertEquals(hashedEnclaveKey, ByteBuffer.wrap(keyDigest))
+
+        // Get the enclave's signature over the random bytes and check its correctness.
```
Where do we check the enclave identity from the attestation?

exFalso replied:
Good point, I added a check that compares the measurement in the enclave binary's metadata and the one in the quote.

exFalso commented on src/main/kotlin/com/r3/sgx/enclave/rng/RngEnclave.kt:
```diff
@@ -0,0 +1,51 @@
+ * The host-enclave communication is set up by the respective sides each building up a pipeline to process messages
+ * flowing back and forth.
+ *
+ * Pipelines consist of a tree of [Handler]s. The root of the tree processes raw ECALLs/OCALLs (calls in and out of the
```
Yeah i wasn't sure about using 'pipeline', but i thought that's the closest thing. I'll change it to 'tree' everywhere. I'll also only use 'root' instead of 'bottom'.

Docs on the rest are in the respective classes, but I'll move some of it here, perhaps it's easier if everything is in one place.

exFalso replied:
@IgorNitto suggested a way to split Handler down a bit further, i'm doing the refactor now. It's conceptually much simpler!

exFalso replied:
Correction: this is punted off for now, as it requires more extensive investigation/experimentation with alternative APIs

exFalso commented on src/main/kotlin/com/r3/sgx/enclave/rng/RngEnclave.kt:
```diff
@@ -0,0 +1,51 @@
+        signatureScheme = api.signatureSchemeFactory.make(SchemesSettings.EDDSA_ED25519_SHA512)
+        keyPair = signatureScheme.generateKeyPair()
+        val keyDigest = MessageDigest.getInstance("SHA-512").digest(keyPair.public.encoded)
+        return Cursor(SgxReportData, ByteBuffer.wrap(keyDigest))
```
It's an `object`, I'll add a comment explaining what it's for

exFalso commented on src/main/kotlin/com/r3/sgx/enclave/rng/RngHandler.kt:
```diff
@@ -0,0 +1,49 @@
+        val publicKey = keyPair.public.encoded
+        // We could use something like protobuf,
+        val size = 4 + randomBytesSize + 4 + publicKey.size + 4 + signature.size
+        connected.sendTopLevel(size) {
```
The serialization/sending is explained in `Sender.kt`. The complexity here stems from efficiency - trying to avoid extra copying.
There is a simpler interface that just sends a ByteBuffer, but that will imply an additional copy when framing the message. But perhaps it's better to use that here for the sake of simplicity?

mikehearn replied:
Yeah in the beginning of this whole space when there's so much to learn, avoiding a single copy probably isn't worth it. The SGX memory authentication adds much more overhead than that. We can always add a hi-perf API later if we find someone who's really bottlenecked on such copies.

exFalso replied:
Ok, I changed it to use BytesHandler with a simpler API

exFalso commented:
@mikehearn @IgorNitto ready for another round

IgorNitto commented:
Looks good, but I have a few questions: 
It is not quite clear if Oblivium's aim is to provide a software development framework or an SGX "app" container (or perhaps both?)? In the former case, we give the software developer the ability to bundle Oblivium API (and SGX platform software) as a dependency of his own application, in order to deploy it and run it on his infrastructure of choice. In the second case, the user builds simply a plugin that needs to be hosted inside an infrastructure that we impose.
In any case, I think this should be made clear, but this sample does not seem to help: i.e., the "how to run it" section does not say how the user is supposed to run the build artifact, which is not a stand-alone application but a Linux dynamic library. Presumably the user is supposed to upload it inside a dedicated hosting server (such as`oblivium/enclavelet-host-server`); perhaps this step should be better illustrated, or maybe at least documented?

exFalso commented:
@IgorNitto The intention is to provide both eventually, i.e. expose the raw building materials for writing enclave and host, and on top of this provide the generic host, containers etc.
In phase 1 it's basically the former, i.e. a JVM-SGX SDK. It should show how to create an enclave and how to write a host. This is kind of ready already, but the documentation needs to be extended to show how to create a fully working setup with release enclaves. We could add the simplified host you're working on now to this phase, we want to encourage using the Enclavelet API anyway to allow uniform handling of enclaves. The reason I punted it off to phase 2 is because with the enclavelet host we would also need to provide a sample host *client* application + documentation + testing which may take a little more time, and I wanted to have an early milestone.
In phase 2 we would provide a sample infrastructure where we host the rng enclaves ourselves, and users can play around with it, get familiar with the chain of trust etc.

mikehearn commented on README.md:
````diff
@@ -0,0 +1,66 @@
+
+To build a properly linked enclave use:
+```
+./gradlew buildEnclaveDebug # Build enclave with debug symbols.
````
Is the only difference symbol stripping or does it affect the SGX debug mode flag too?

exFalso replied:
It doesn't affect the debug mode flag, it could in theory be loaded in non-debug mode as long as it's signed and the config xml disables debug mode

mikehearn commented on src/test/kotlin/com/r3/sgx/enclave/rng/RngEnclaveTest.kt:
```diff
@@ -0,0 +1,154 @@
+        // - Whether you want to produce linkable or unlinkable quotes
+        // - Your SPID
+        //
+        // "Linkability" relates to the EPID provisioning protocol. This a blind join protocol used to create the
```
Typo: This is

mikehearn commented on src/test/kotlin/com/r3/sgx/enclave/rng/RngEnclaveTest.kt:
```diff
@@ -0,0 +1,154 @@
+        // See https://software.intel.com/en-us/articles/certificate-requirements-for-intel-attestation-services for
+        // further details.
+        //
+        // In this test we use dummy values as we are not doing a fu.
```
doing a fu?

exFalso commented:
@IgorNitto I created a JIRA for adding a minimal host app in a followup PR
@mikehearn Addressed comments. The muxing is done in MuxingHandler, used by the Channel handlers and the Root handler (see oblivium repo for source)
